crypto/hash/sha3.go

```diff
@@ -1,65 +1,193 @@
 package hash
 
-import (
-	"hash"
-
-	"golang.org/x/crypto/sha3"
+const (
+	rateSha3_256 = 136
+	rateSha3_384 = 104
 )
 
-// sha3_256Algo, embeds commonHasher
-type sha3_256Algo struct {
-	*commonHasher
-	hash.Hash
+// NewSHA3_256 returns a new instance of SHA3-256 hasher.
+func NewSHA3_256() Hasher {
+	return &sha3State{
+		rate:      rateSha3_256,
+		outputLen: HashLenSha3_256,
+	}
 }
 
-// NewSHA3_256 returns a new instance of SHA3-256 hasher
-func NewSHA3_256() Hasher {
-	return &sha3_256Algo{
-		commonHasher: &commonHasher{
-			algo:       SHA3_256,
-			outputSize: HashLenSha3_256},
-		Hash: sha3.New256()}
+// NewSHA3_384 returns a new instance of SHA3-384 hasher.
+func NewSHA3_384() Hasher {
+	return &sha3State{
+		rate:      rateSha3_384,
+		outputLen: HashLenSha3_384,
+	}
+}
+
+// Size returns the output size of the hash function in bytes.
+func (d *sha3State) Size() int {
+	return d.outputLen
 }
 
-// ComputeHash calculates and returns the SHA3-256 output of input byte array.
-// It does not reset the state to allow further writing.
-func (s *sha3_256Algo) ComputeHash(data []byte) Hash {
+// Algorithm returns the hashing algorithm of the instance.
+func (s *sha3State) Algorithm() HashingAlgorithm {
+	switch s.outputLen {
+	case HashLenSha3_256:
+		return SHA3_256
+	case HashLenSha3_384:
+		return SHA3_384
+	}
+	return UnknownHashingAlgorithm
+}
+
+// ComputeHash calculates and returns the SHA3 digest of the input.
+// It updates the state and doesn't allow further writing without
+// calling Reset().
+func (s *sha3State) ComputeHash(data []byte) Hash {
 	s.Reset()
 	_, _ = s.Write(data)
```

Comment:
I would be almost tempted to change https://github.com/onflow/flow-go/blob/dd7a60344577faf159bfec0ec55c32bcdc4bf62e/crypto/hash/sha3.go#L131
to 
```golang
func (d *sha3State) Write(p []byte) (written int, err error) {
    return d.write(p), nil
}

func (d *sha3State) write(p []byte) written int {
```
and call `s.write(data)` here ...